### PR TITLE
fix/#36-ex-post-calculation-grid-investment

### LINKED ISSUE #36: add ex-post calculations for the grid
Calculate the following numbers for the grid

- [x] Investitionen Leitungen #32
- [x] Netzausbau je Line #32
- [x] Netzauslastung nach Netzausbau, relativ je Line
- [x] Netzverluste in MWh
- [x] CO2 Emissionen

Flows on the lines are [already extracted](https://github.com/windnode/WindNODE_ABW/blob/ca3a1fb58391b00955e7f3f5707585802a116171/windnode_abw/analysis/tools.py#L565-L570) from the results.

### PR DESCRIPTION
**Pull request by gplssm**
Fix #36

### CODE REVIEW

**Comment by gplssm** [comment]
Some notes on what was done and how

- costs and CO2 emissions calculated for lines (existing and new separately)
- costs on links where only accounted once, the max. capacity of both directions is taken for this
- to use the new code, you need a new run because more results are extracted which happens in `run_scenario.py`
- newly installed capacity is considered in line loading

Two questions remain from my side:

1. I saw in model.py that opex_var aren't multiplied by length, annuity is. Is this correct?
2. How are `emissions_fix` defined? The implementation right now assumes that these are g/kW

**Comment by nesnoj** [comment]
Note: I haven't checked the final results, this needs to be done in #76.

File: windnode_abw/analysis/tools.py [hunk 1]
```diff
@@ -481,6 +487,21 @@ def extract_flow_params(flow_params_raw, node_pattern, bus_pattern, stubname,
...
+
+    # Get an extract of relevant data
+    vars_extract = vars.loc[
+                    vars.index.get_level_values(0).str.match(node_pattern),
+                    vars.index.get_level_values(1).str.match(bus_pattern), :].astype(float)
+
+    # transform to wide-to-long format while dropping bus column level
+    vars_extract = vars_extract.droplevel(1).groupby(level=0).max()
+    vars_extract.index = pd.MultiIndex.from_frame(vars_extract.index.str.extract(node_pattern))
+    vars_extract = vars_extract.sum(level=list(range(vars_extract.index.nlevels)))
```
[the remark this answers is not in the document]
Comment (gplssm) [reply windnode_abw/analysis/tools.py:500]: These lines assign a new multilevel-index to the extracted data which is based on a certain split of the flow label describe by `node_pattern`. In addition, and specifically in the case of grid investments, it is intended to take the maximum value of two flow investments representing the same line, but in different direction (L 497). It's actually working correctly, but L 497 could be rewritte like `vars_extract = vars_extract.max(level=0)`.

Any further concerns?

File: windnode_abw/analysis/tools.py [hunk 2]
```diff
@@ -956,6 +980,18 @@ def _extract_tech_params(NAMES):
...
     params["Installed capacity grid"] = _rename_external_hv_buses(line_capacity, region, merged=True)[0]
 
+    # Newly installed capacity grid
+    params["Newly installed capacity grid per bus"] = extract_invest(
+        results_raw["invest"],
+        "line_(?P<line_id>\d+)_b(?P<bus_from>\d+)_b(?P<bus_to>\d+)",
+        'b_el_\d+')
+    params["Newly installed capacity grid"] = _rename_external_hv_buses(
+        params["Newly installed capacity grid per bus"].copy(),
+        region, merged=True)[0]
```
[the remark this answers is not in the document]
Comment (gplssm) [reply windnode_abw/analysis/tools.py:990]: Can't confirm this using `run_timestamp = '2020-07-22_063122_fullyear_#75'` and `scenarios = ['ISE2050_Relimit']`
grafik

:thinking:

File: windnode_abw/analysis/tools.py [hunk 3]
```diff
@@ -1011,6 +1050,8 @@ def _calculate_supply_costs(capacity, generation, params):
     results["Wärmespeicher nach Gemeinde"] = extracted_results["Wärmespeicher"].sum(level=["level", "ags"])
     results["Batteriespeicher nach Gemeinde"] = extracted_results["Batteriespeicher"].sum(level=["level", "ags"])
     results["Stromnetzleitungen"] = extracted_results["Stromnetz"].sum(level=["ags_from", "ags_to"])
+    results["Stromnetzleitungen per bus"] = extracted_results["Stromnetz per bus"].sum(
+        level=["line_id", "bus_from", "bus_to"])
```
[the remark this answers is not in the document]
Comment (gplssm) [reply windnode_abw/analysis/tools.py:1054]: The negative sign might be confusing...
>0: flow from `from_bus` to `to_bus`
<0: flow from `to_bus` to `from_bus`
0: no flow

> Neg. values lead to neg. emissions in co2 calc below. As var. emissions are 0 this currently doesn't matter but will blow up the results, if this value is changed!
True! Could become a problem. Changed it with `abs()`

[the remark this answers is not in the document]
Comment (gplssm) [reply windnode_abw/analysis/tools.py:1054]: > By the way, in `_calculate_co2_emissions()` you use the params from tech_assumptions, right? Doesn't the line length need to be incorporated here? I think so..
> 
> Same for opex var in `_calculate_supply_costs()` (though it's currently 0).
> 
> I'm not fully sure, please recheck..

True, partially, actually it depends how the `opex_var` are defined. In `EUR/MWh` or in `EUR/MWh/km` :smile: I wouldn't change it though

File: windnode_abw/analysis/tools.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (nesnoj) [reply windnode_abw/analysis/tools.py:500]: sorry, this was too unspecific. I meant L 497.

File: windnode_abw/analysis/tools.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (nesnoj) [reply windnode_abw/analysis/tools.py:990]: Hmm, same values at my end. Dunno what was wrong before..

File: windnode_abw/analysis/tools.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (nesnoj) [reply windnode_abw/analysis/tools.py:1054]: > True, partially, actually it depends how the opex_var are defined. In EUR/MWh or in EUR/MWh/km smile I wouldn't change it though

It's EUR/MWh/km.
Ok, I added a note to the tech ods not to change :skull: .

**Comment by gplssm** [comment]
> Thanks for this hot new stuff!
> 
> I tested with the full year run of ISE2050_Relimit (`2020-07-22_063122_fullyear_#75.tar.gz` on the server) and have some remarks:
> 
>     1. There are investments for EHV/HV trafos included in invest.csv which you do not analyze/add to the invest costs. As we've only 2 of them and the cap. is a best guess (from ego) I'm fine with neglecting it.

We could open an issue for the nice-to-have list for that. You decide!

>     2. The line investments are done continuously and so are the emissions (in contrast to real grid expansion). I better won't open up this can of works here, right? ;)

:arrow_right: Even-nicer-to-have list

>     3. Investment approach: Will the investment (annuity/EPC) be made per direction (twice) for each Link? I'm not talking about only using the max of both (as you correctly did in the results extraction!) but I think the double investments as part of the optimization problem may "demotivate" grid expansion?
>        This is more a general oemof questing we're not going to fix here, but I think it's important enough to be kept in mind.

Yepp, it is accounted twice

I've used your [minimum example](https://gist.github.com/nesnoj/9c31238f104836734aa62bdaf65c99ae) and saw this

```
\* Source Pyomo model name=Model *\

min 
objective:
+100 InvestmentFlow_invest(line_0_b_0)
+100 InvestmentFlow_invest(line_0_b_1)
+25000 ONE_VAR_CONSTANT

s.t.

c_e_Bus_balance(b_0_0)_:
-1 flow(b_0_line_0_0)
+1 flow(line_0_b_0_0)
= -300
```

You're totally right, twice
1. This is simply wrong
2. We won't fix it here

Shall we file an issue for oemof-solph?

>     4. Rather general: I'm a little curious that `params_stat.csv` is not used in the repo code at all, where are static params (e.g. `nominal_storage_capacity` of storages) are obtained from? Aren't they needed for the analysis of flex. results?

I think I get these parameters from `region.tech_assumptions` and other attributes of `region`. So far, we haven't used to storage fill level `nominal_storage_capacity` in our analysis.

> Also, find comments on the code.
> 
> > Two questions remain from my side:
> > 
> > 1. I saw in model.py that opex_var aren't multiplied by length, annuity is. Is this correct?
> > 2. How are emissions_fix defined? The implementation right now assumes that these are g/kW
> 
>     1. Currently, this does not matter, as it is 0 in the current settings (only opex_fix>0) but in general you're right, thanks! Same applies for var emissions (also currently 0). I fixed both.
> 
>     2. No, `g/(kW*km)` resp. `kg/(MW*km)`. You have to use the length, see code comment on this.

Ok. Fix it - somehow

**Comment by nesnoj** [comment]
Note @nailend: As soon as this one is merged, use new runs
`2020-07-24_054949_1month_#75` or `2020-07-22_063122_fullyear_#75`
for analysis due to changed file format.

**Comment by nesnoj** [comment]
:partying_face:

**Pull request merged**